compiler/z/codegen/ConstantDataSnippet.cpp

```diff
@@ -406,71 +406,81 @@ uint8_t *
 //
-//  0x00000100211cfed0     DC  0x00000000       ; Cached Class4
-//  0x00000100211cfed8     DC  0x00000000       ; Cached Method4
+//  DC  0x00000000       ; Cached Class4   (unused if the target is direct)
+//  DC  0x00000000       ; Cached Method4  (unused if the target is direct)
 //
 //  and so on...
 
 // If we use CLFI / BRCL
-//  0x00000100211cfea0     DC  0x00000000       ; Patch slot of the first CLFI (Class 1)
-//  0x00000100211cfea0     DC  0x00000000       ; Patch slot of the second CLFI (Class 2)
-//  0x00000100211cfea0     DC  0x00000000       ; Patch slot of the third CLFI (Class 3)
+//  DC  0x00000000       ; Patch slot of the first CLFI (Class 1)
+//  DC  0x00000000       ; Patch slot of the second CLFI (Class 2)
+//  DC  0x00000000       ; Patch slot of the third CLFI (Class 3)
 // and so on ...
 
 //  31-Bit Layout.
-//  0x00000000211cfe50     DC  0x0000610ff8     ; Address Of Dispatch Glue
-//  0x00000000211cfe54     DC  0x00211cf58a     ; Call Site RA
-//  0x00000000211cfe58     DC  0x2020b380       ; Address Of Constant Pool
-//  0x00000000211cfe5c     DC  0x00000040       ; CP Index
-//  0x00000000211cfe60     DC  0x00000000       ; Interface class
-//  0x00000000211cfe64     DC  0x00000000       ; Method Index
+//  DC  0x0000610ff8     ; Address Of Dispatch Glue
+//  DC  0x00211cf58a     ; Call Site RA
+//  DC  0x2020b380       ; Address Of Constant Pool
+//  DC  0x00000040       ; CP Index
+//  DC  0x00000000       ; Interface class
+//  DC  0x00000000       ; Method Index
+//  DC  0x00000000       ; Thunk address
+//  DC  0x00000000       ; Cached Class1
+//  DC  0x00000000       ; Cached Method1
 //
-//  0x00000000211cfe68     DC  0x00000000       ; Cached Class1
-//  0x00000000211cfe6c     DC  0x00000000       ; Cached Method1
-//
-//  0x00000000211cfe70     DC  0x00000000       ; Flags (resolved)
+//  DC  0x00000000       ; Flags (resolved)
 
    TR::Compilation *comp = cg()->comp();
    int32_t i = 0;
    uint8_t * cursor = cg()->getBinaryBufferCursor();
    AOTcgDiag1(comp, "TR::S390InterfaceCallDataSnippet::emitSnippetBody cursor=%x\n", cursor);
 
    // Class Pointer must be double word aligned.
-   int32_t padBytes = ((intptrj_t)cursor + 5 * sizeof(intptrj_t) + 7) / 8 * 8 - ((intptrj_t)cursor + 5 * sizeof(intptrj_t));
-   cursor += padBytes;
+   // For 64-bit single dynamic slot, LPQ and STPQ instructions will be emitted; and they require quadword alignment
+   //
+   // 4-slot snippet has 10 DC fields before the 1st class pointer; whereas single dynamic snippet has 6.
+   // To align class pointers to double/quad word boundaries, we can just align the cursor to boundaries instead
+   // of aligning the the first PIC.
+   int32_t alignment = comp->getOption(TR_enableInterfaceCallCachingSingleDynamicSlot) ? 16 : 8;
+   if (((intptrj_t)cursor % alignment) != 0)
+      {
+      cursor += (alignment - ((intptrj_t)cursor % alignment));
+      }
```

Comment:
Could you please add a fatal assert following this alignment to validate that `cursor` is aligned properly? This will guard against future modifications to the snippet layout misaligning things.

Reply:
Actually I see there is one below but it only checks for 8 byte alignment? Don't we need 16 byte?

Reply:
I'll update the assert to handle 16-byte alignment check.